wallet-api/src/Ledger/Types.hs

```diff
@@ -223,23 +223,44 @@ deriving anyclass instance ToJSON Address
 deriving anyclass instance FromJSON Address
 
 -- | Script
-newtype Script = Script { getSerialized :: BSL.ByteString }
-  deriving newtype (Serialise, Eq, Ord)
+newtype Script = Script { getPlc :: PLC.Program PLC.TyName PLC.Name () }
+  deriving newtype (Serialise)
+
+{- Note [Eq and Ord for Scripts]
+We need `Eq` and `Ord` instances for `Script`s mostly so we can put them in `Set`s.
+However, the `Eq` instance for `Program`s is *alpha-equivalence*, and we don't
+have a compatible `Ord` instance, nor is it easy to derive one.
+
+So we piggyback off a different representation. In this instance we have two
+options:
+- Use the serialized form
+- Use a hash
+The problem with the latter is that we don't want to add a derived `Hashable` instance
+for `Program`s that's not compatible with the `Eq` instance. We *could* add a derived
+instance for `Program`s with de Bruijn indices, since in that case the derived `Eq`
+coincides with alpha-equivalence. However, this might be faster.
+
+For the moment we use the serialized form. We used to store the serialized form directly
+in `Script`, but that led to a lot of deserializing and reserializing in `applyProgram`.
+Here we have to serialize when we do `Eq` or `Ord` operations, but this happens comparatively
+infrequently (I believe).
+-}
+instance Eq Script where
+    a == b = serialise a == serialise b
```

Comment:
Why not compare using the `Eq` instance of `Program` directly?

Reply:
Then it won't be compatible with the `Ord` instance, since it will say some things are equal which do not have `compare x y = EQ`.

Reply:
I see, thanks.